// zayra-baileys-lib/index.js
const { makeWASocket, useMultiFileAuthState, DisconnectReason } = require('@whiskeysockets/baileys');

async function connectDctBot() {
    const { state, saveCreds } = await useMultiFileAuthState('auth_info_dct');

    const sock = makeWASocket({
        auth: state,
        printQRInTerminal: true
    });

    sock.ev.on('connection.update', (update) => {
        const { connection, lastDisconnect } = update;
        if (connection === 'close') {
            const shouldReconnect = lastDisconnect?.error?.output?.statusCode !== DisconnectReason.loggedOut;
            if (shouldReconnect) connectDctBot();
        } else if (connection === 'open') {
            console.log('✅ dct Baileys connected!');
        }
    });

    sock.ev.on('creds.update', saveCreds);

    // Message & Button handler
    sock.ev.on('messages.upsert', async (m) => {
        const msg = m.messages[0];
        if (!msg.key.fromMe && m.type === 'notify') {
            const from = msg.key.remoteJid;
            const text = msg.message?.conversation || '';

            const selectedButton = msg.message?.buttonsResponseMessage?.selectedButtonId;
            if (selectedButton) {
                if (selectedButton === 'btn_yes') {
                    await sock.sendMessage(from, { text: 'Thank you! 😊' });
                } else if (selectedButton === 'btn_no') {
                    await sock.sendMessage(from, { text: 'Okay!' });
                }
                return;
            }

            if (text.toLowerCase() === 'button') {
                await sendButtons(sock, from);
            }
        }
    });

    return sock;
}

// ====================== PAIRING CODE FUNCTION ======================
async function connectWithPairingCode(phoneNumber) {
    const { state, saveCreds } = await useMultiFileAuthState('auth_info_dct');

    const sock = makeWASocket({
        auth: state,
        printQRInTerminal: false   // Pairing code use කරනකොට QR print කරන්න එපා
    });

    sock.ev.on('connection.update', (update) => {
        const { connection, lastDisconnect } = update;
        if (connection === 'close') {
            const shouldReconnect = lastDisconnect?.error?.output?.statusCode !== DisconnectReason.loggedOut;
            if (shouldReconnect) connectWithPairingCode(phoneNumber);
        } else if (connection === 'open') {
            console.log('✅ Connected using Pairing Code!');
        }
    });

    sock.ev.on('creds.update', saveCreds);

    // Pairing code එක generate කරන එක
    if (!sock.authState.creds.registered) {
        const code = await sock.requestPairingCode(phoneNumber);
        console.log(`\n📱 Your Pairing Code: ${code}`);
        console.log('Go to WhatsApp → Linked Devices → Link with phone number → Enter this code\n');
    }

    return sock;
}

// ====================== SEND BUTTONS ======================
async function sendButtons(sock, jid) {
    const buttons = [
        { buttonId: 'btn_yes', buttonText: { displayText: 'Yes ✅' }, type: 1 },
        { buttonId: 'btn_no', buttonText: { displayText: 'No ❌' }, type: 1 }
    ];

    await sock.sendMessage(jid, {
        text: 'Do you like this bot?',
        buttons: buttons,
        headerType: 1
    });
}

module.exports = { connectDctBot, connectWithPairingCode, sendButtons };
